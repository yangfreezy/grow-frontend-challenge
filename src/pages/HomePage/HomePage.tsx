import React from 'react';
import { useState } from 'react';
import { Container } from '@mui/system';

import useLocalStorageState from '../../hooks/useLocalStorageState';
import { WikiPageViewMetrics } from '../../types/wiki';
import { getArticleId } from '../../utils/wiki';
import ResultsFilters from './components/ResultsFilters';
import PinnedResultsTable from './components/ResultsTablePinned';
import ResultsTable from './components/ResultsTable';
import WikiPageHeader from './components/WikiPageHeader';
import moment from 'moment';

// ADD a loading animation for the api calls
// Add something for when there are no results


const pageLayoutStyles = {
  display: 'flex', alignItems: 'center', flex: 1, flexDirection: 'column'
};

const HomePage = () => {
  const [pinnedArticles, setPinnedArticles] = useLocalStorageState([], 'pinnedArticles');
  const [results, setResults] = useState<WikiPageViewMetrics[]>([]);

  const getIsArticleSelected = (result : WikiPageViewMetrics, prev: WikiPageViewMetrics[] = pinnedArticles) : boolean => {
    const uniqueArticleId = getArticleId(result);
    if (prev.some(a => getArticleId(a) === uniqueArticleId)) return true;
    return false;
  }

  const removePinnedArticle = (id: string) => pinnedArticles.filter((a : WikiPageViewMetrics) => getArticleId(a) !== id)

  const addPinnedArticle = (article: WikiPageViewMetrics) => {
    // Adding a timestamp since this article can persist multiple days and the rank will otherwise lose that context
    return pinnedArticles.concat([{ ...article, timestamp: moment().format("MMM Do, YYYY") }]);
  }

  const togglePinnedArticles = (result : WikiPageViewMetrics) => {
    setPinnedArticles((prev: WikiPageViewMetrics[]) => {
      const alreadySelected = getIsArticleSelected(result, prev);
      const uniqueArticleId = getArticleId(result);
      return alreadySelected ? removePinnedArticle(uniqueArticleId) : addPinnedArticle(result);
    });
  }

  const showPinnedArticles = pinnedArticles?.length > 0;

  return (
    <Container sx={pageLayoutStyles}>
      <WikiPageHeader />
      <ResultsFilters setResults={setResults} />
      <>
        {
          showPinnedArticles ? (
            <PinnedResultsTable
              pinnedArticles={pinnedArticles}
              togglePinnedArticles={togglePinnedArticles}
            />
          ) : null
        }
        <ResultsTable
          results={results}
          getIsArticleSelected={getIsArticleSelected}
          pinnedArticles={pinnedArticles}
          togglePinnedArticles={togglePinnedArticles}
        />
      </>
    </Container>
  )
};

export default HomePage;
