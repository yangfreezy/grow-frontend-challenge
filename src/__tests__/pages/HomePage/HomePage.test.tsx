import React from "react";
import HomePage from "../../../pages/HomePage/HomePage";
import { render, RenderResult } from '@testing-library/react';

describe('HomePage', () => {
    let app: RenderResult;
    beforeEach(() => {
        app = render(<HomePage/>);
    })
    it('renders', () => {
        expect(app).toBeTruthy();
    })

    it('matches the snapshot', () => {
    expect(app).toMatchSnapshot();
    })

    it('renders the header', () => {
        expect(app.getByTestId('WikiPageHeader')).toBeTruthy();
    })

    it('renders the results filters', () => {
        expect(app.getByTestId('ResultsFilters')).toBeTruthy();
    })

    it('does not render the pinned results table initially', () => {
        expect(app.queryByTestId('PinnedResultsTable')).toBeFalsy();
    })

    it('renders the results table', () => {
        expect(app.getByTestId('ResultsTable')).toBeTruthy();
    })

});
