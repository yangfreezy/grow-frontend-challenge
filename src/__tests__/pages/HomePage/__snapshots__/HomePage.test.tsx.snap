// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage matches the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-crdix7-MuiContainer-root"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 css-1tqoyhh-MuiTypography-root"
          data-testid="WikiPageHeader"
        >
          Top 
          <b>
            Wiki
          </b>
           Articles
        </h2>
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-1mxi8ur-MuiContainer-root"
          data-testid="ResultsFilters"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r2:"
              id=":r2:-label"
            >
              Date
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-readOnly css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                aria-label="Choose date, selected date is Dec 14, 2022"
                aria-readonly="true"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-readOnly css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r2:"
                readonly=""
                type="text"
                value="12/14/2022"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Date
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root css-1km2w0t-MuiFormControl-root"
            data-testid="CountrySelector"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              id="country-quantity-helper-label"
            >
              Country
            </label>
            <div
              aria-label="Country Selector"
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              data-testid="CountrySelectorSelect"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="country-quantity-select-label country-quantity-select"
                class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                id="country-quantity-select"
                role="button"
                tabindex="0"
              >
                United States
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="US"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Country
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              id="results-quantity-helper-label"
            >
              Results
            </label>
            <div
              aria-label="Number of Results Selector"
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-labelledby="results-quantity-select-label results-quantity-select"
                class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                id="results-quantity-select"
                role="button"
                tabindex="0"
              >
                100
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="100"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Results
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <table
          class="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
          data-testid="ResultsTable"
        >
          <thead
            class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignLeft MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
              >
                ðŸ“Œ
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignLeft MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
                scope="col"
              >
                Rank
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignRight MuiTableCell-sizeMedium css-i02g0k-MuiTableCell-root"
                scope="col"
              >
                Article
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignRight MuiTableCell-sizeMedium css-i02g0k-MuiTableCell-root"
                scope="col"
              >
                Views
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          />
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-crdix7-MuiContainer-root"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 css-1tqoyhh-MuiTypography-root"
        data-testid="WikiPageHeader"
      >
        Top 
        <b>
          Wiki
        </b>
         Articles
      </h2>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1mxi8ur-MuiContainer-root"
        data-testid="ResultsFilters"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for=":r2:"
            id=":r2:-label"
          >
            Date
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-readOnly css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              aria-label="Choose date, selected date is Dec 14, 2022"
              aria-readonly="true"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-readOnly css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r2:"
              readonly=""
              type="text"
              value="12/14/2022"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Date
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root css-1km2w0t-MuiFormControl-root"
          data-testid="CountrySelector"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            id="country-quantity-helper-label"
          >
            Country
          </label>
          <div
            aria-label="Country Selector"
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
            data-testid="CountrySelectorSelect"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="country-quantity-select-label country-quantity-select"
              class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="country-quantity-select"
              role="button"
              tabindex="0"
            >
              United States
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value="US"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Country
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            id="results-quantity-helper-label"
          >
            Results
          </label>
          <div
            aria-label="Number of Results Selector"
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          >
            <div
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby="results-quantity-select-label results-quantity-select"
              class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id="results-quantity-select"
              role="button"
              tabindex="0"
            >
              100
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value="100"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Results
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <table
        class="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
        data-testid="ResultsTable"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignLeft MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
            >
              ðŸ“Œ
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignLeft MuiTableCell-sizeMedium css-y8ay40-MuiTableCell-root"
              scope="col"
            >
              Rank
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignRight MuiTableCell-sizeMedium css-i02g0k-MuiTableCell-root"
              scope="col"
            >
              Article
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-alignRight MuiTableCell-sizeMedium css-i02g0k-MuiTableCell-root"
              scope="col"
            >
              Views
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        />
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
